File: backend/app/schemas/pipeline.py
from datetime import datetime

from pydantic import BaseModel

from app.schemas.queue import QueueResponse
from app.schemas.track import TrackResponse
from app.schemas.track_summary import TrackStatusSummary


class PipelineEvent(BaseModel):
    message: str
    at: datetime


class PipelineSnapshot(BaseModel):
    queue: QueueResponse
    tracks: list[TrackResponse]
    track_summary: TrackStatusSummary
    worker_active: bool
    queue_stalled: bool = False
    queue_backlogged: bool = False
    last_event: PipelineEvent | None = None

File: frontend/src/types/pipeline.ts
import type { QueueSummary } from './queue'
import type { Track } from './track'

export interface TrackStatusSummary {
  total: number
  by_status: Record<string, number>
}

export interface PipelineEvent {
  message: string
  at: string
}

export interface PipelineSnapshot {
  queue: QueueSummary
  tracks: Track[]
  track_summary: TrackStatusSummary
  worker_active: boolean
  queue_stalled: boolean
  queue_backlogged: boolean
  last_event: PipelineEvent | null
}

export interface PipelineWsMessage {
  type: 'snapshot'
  message?: string | null
  data: PipelineSnapshot
}
